### ISSUE TITLE
package layout

### ISSUE THREAD

**Issue by kieranjmartin** [pharmaR/riskmetric#2]
Hi,

I have a few suggestions about how this package should be laid out. Currently the packages info is added in one tibble, but this is a bit hard to future proof, and makes the code harder to read. Also the primary function of the risk metrics code is to generate an Rd file. I think the best way to generate package data is to store the code doing so in data-raw, following Hadley Wickham's advice [here](http://r-pkgs.had.co.nz/data.html) .

So I'd have something like

```
package_details <- list()

package_details$dplyr <- tibble::tibble(package = "dplyr", source = "CRAN", `21CFR` = "N")

package_details$emmeans <- tibble::tibble(package = "dplyr", source = "CRAN", `21CFR` = "N")

package_tibble <- purrr::reduce(package_details, dplyr::bind_rows)

usethis::use_data(package_tibble, overwrite = TRUE)
```
which makes an rdata file in /data . You can then make an R file called package_tibble which documents the data.

I'd then have a function called risk_metrics(packagedata = package_tibble) which takes the inputted tibble and gives up to date metrics. So I'd put stuff that is static and determined by us (package, location, 21cfr, other measures) in data_raw, then have functions to determine things like code coverage by making external calls.

For an intermediate point, we could simply hard code functions, so

```
code_cov(packagename)<- function{

if (packagename == "dplyr") return(83)
}
```

This then gives the user flexibility to add a new package to the package_tibble with details, and get information using the functions we've built.

If you like, I can make a fork implementing something like this.

**Comment by dgkf** [comment]
Thanks, @kieranjmartin - all great suggestions.

We discussed in our meeting yesterday that the current `tibble` format is largely for communicating with the rest of the team about what information we hope to derive automatically and to gather feedback about what more might be needed.

In the long run, you’re absolutely right that a function to extract these metrics will be needed. 

We’re currently targeting the end of March, and doing a fairly manual pass at the process seemed to be the most achievable compromise that we could squeeze into that timeframe. Beyond that, I expect we'll have more flexibility to look longer-term.

**Comment by elong0527** [comment]
Thanks @kieranjmartin, I added you as a contributor. Feel free to create a brunch and work on it. 

Some bullet of thoughts:

1. I like the listing structure in the long run that easily handle the unstructured data nicely. The current structure is just for exploration purpose. 

2. I start with `usedata` function, yet struggle on the `Rdata` vs `Rd` difference. As I expect riskmetric is a single object, I chose to save in `Rd` file and open for discussion. In the long run, I expect a function like `tools::clean_CRAN_db` is created to gather data from a server that automatically maintain risk metric. (So no data need to be saved in the R package)

3. `version` should be taken into consideration too, if our final plan is to derive the risk score for every version of a R package on CRAN/Bioconductor.

**Comment by kkmann** [comment]
Hi everyone,

two points on the technical side of things

1. Wouldn't it be easier to start by implementing each metric in its own function? Way easier to maintain and much more modular. People can work on stuff independently.

2. Are you sure that you want to go through the trouble of creating one package encompassing both Bioconductor and CRAN? These two are so completely separate (and potentially different metrics are of interest to the different user communities of those?) that it would be much easier to split it in a CRAN / BioConductor package each. (saferCRAN, saferBioConductor ?) Just getting a package with BioConductor dependencies on CRAN seems daunting...

@dgkf what do you think?

**Comment by kkmann** [comment]
Another interesting question is the input to function `get_metric_xyz()`, could be 

1. package name as character -> scrape known repositories for information
2. repository -> quite a few packages are still not developed in a public repository...
3. source package itself (.tar.gz) -> most information, can parse README.md and scrape standard repositories

I guess 3) would be the most robust/widely applicable method, right?

**Comment by dgkf** [comment]
Thanks for all the thoughtful considerations, @kkmann - I'll try to address the suggestions


> Wouldn't it be easier to start by implementing each metric in its own function?

Absolutely. I don't think the current structure scales particularly well when adding new metrics - nor was it intended to be a lasting solution. In the long run, I see us having a function like `assess_package('pkgname')` (or `assess_package('pkg/install_or_source/path')`) which will then iterate through a list of metric derivation functions.


> it would be much easier to split it in a CRAN / BioConductor package each

I'm pretty unfamiliar with the ramifications of mixing dependency sources - I'd have to defer to someone more familiar to weigh in. As a design goal, I think it's important to have some alignment between CRAN and Bioconductor package metrics. If it ends up they can't be reconciled we could consider separate metrics.


> Another interesting question is the input to function get_metric_xyz()

In the long run we probably need an extensible mechanism of deciding which route to choose. Perhaps `get_metrics_xyz()` style functions should accept a parameter to specify preference of source such as `get_metrics_xyz(<pkg>, try_sources = c("repo", "targz", "scraped"))`.

As we get started, we can rewrite our current code such that we only run each function with `try_sources = "scraped"` (i.e. "scraped from CRAN / Bioconductor"), and we can slowly add in more robust functionality to derive each metric from the installed `tar.gz` or after cloning the repo.

**Comment by elong0527** [comment]
@kieranjmartin and @dgkf , totally agree. The original function is just for proof of concept. It will be easier to maintain by having one `assess_packages` function that call multiple `get_metrics` functions. Maybe we could work in parallel to achieve the targets below for the month of April? 

* Save the expanded R package listing and source in a csv format and put it in the ./data folder
* Create the function architectures specification. (we may keep the body of function in blank and just put it in the R folder as a spec.   

After we agreed with the structure, we can improve the functionality of each function.

**Comment by kkmann** [comment]
One more random thought - I am a big fan of starting small and then adding features. I guess (correct me if I am wrong) that the main focus will be CRAN packages since CRAN itself can be seen as a way of achieving QC. We should therefore probably start by focussing on CRAN and maybe extend to BioConductor later. Maybe we can get around a strong dependency on BC by using the 'Suggests' field to declare an optional dependency on BC and some BC packages. Alternatively, we just create an 'add-on' bioconductor package extending the function of the core CRAN-based package - but we should play around with that and decide it later if we agree that the initial focus will be CRANtatistic packages.

In that case the first step would be to fix the CI build and remove the bioconductor dependencies. I did not want to open a PR since this seems to be a rather fundamental decision - @elong0527?

Also sifted through the packages alredy out there (packagemetrics), seem to be quite dead. I would suggest taking what we need there (exceellent code base! properly citing of course) to avoid dependency hell and get full control.

**Comment by dgkf** [comment]
Regarding `packagemetrics`, @bzkrouse is one of the contributors to `packagemetrics` and is actively involved in the riskmetric work. @bzkrouse - do you have any insight into how we may want to proceed as we consider making `packagemetrics` a dependency?

I know we've discussed at the very least taking inspiration from `packagemetrics`, but it might be good to decide at some point whether or not it's worth drawing the dependency.

My personal preference is to greedily include dependencies early on. If we decide something isn't worth the added overhead, we can always take in the needed functionality later, but it helps to minimize the code base as we're co-developing in the prototype stages. If we go down that route, I usually maintain little `# TODO` reminders in places where I know I'll eventually want to strip out a dependency.

**Comment by kkmann** [comment]
:+1:, I'd strongly suggest that we remove the dependency on bioconductor for now, just to get things rolling. I'd be happy to open a PR to do so, but it would essentially be an entire overhaul of what we have now, so it only makes sense if you guys are happy with it.
My point is that we should aim for CRAN conformity from the very beginning - would be odd to me if the riskmetric package itself wasn't CRAN-compatible ;)

**Comment by elong0527** [comment]
I think it is good to start from CRAN. That is already a big win If we could have a consensus with agency for using a group of “mature” tidyverse R packages for regulatory works by the end of this year. 

@kkmann  I have give you the write access for this repo. Yeah, let’s discuss the details in a PR request you created.

**Comment by bzkrouse** [comment]
> Regarding `packagemetrics`, @bzkrouse is one of the contributors to `packagemetrics` and is actively involved in the riskmetric work. @bzkrouse - do you have any insight into how we may want to proceed as we consider making `packagemetrics` a dependency?
> 
> I know we've discussed at the very least taking inspiration from `packagemetrics`, but it might be good to decide at some point whether or not it's worth drawing the dependency.
> 
> My personal preference is to greedily include dependencies early on. If we decide something isn't worth the added overhead, we can always take in the needed functionality later, but it helps to minimize the code base as we're co-developing in the prototype stages. If we go down that route, I usually maintain little `# TODO` reminders in places where I know I'll eventually want to strip out a dependency.

@dgkf - This makes a lot of sense.  While it doesn't make sense to reinvent the wheel during the prototype stage, it is important to build the right long-term infrastructure for our own needs.  I'll circle back to the packagemetrics team to provide a status update get more direction about the future of the package. 

Also, this thread is full of great ideas.  Focusing on CRAN packages sounds like a great starting point.

**Comment by kkmann** [comment]
Cool! So, with @elong0527 taking care of setting up the CI/CD infrastructure + pkdown I think we are ready to go. 

1) How do we organize communication between slack/GitHub issues? I am pretty much in favour of moving as much as possible to GitHub issues since this would mean that everything is in one place.
It is also quite difficult to separate the conceptual from the technical side of things anyway ;)

2) We should try to maintain a fairly up-to-date vision statement in the index.Rmd for our pkgdown site. This should summarize high-level ideas and give a concise description of what we try to achieve.

3) Depending on how many others want to contribute, contribution guidelines might be a good idea. Not everyone will be familiar with a protected master branch workflow and code reviews, right?

4) How do we orchestrate work across the different metrics? Since each metric will go in its own file these can be worked on pretty independently. May we have a 'metrics ideas' issues for people to call dibs on metrics etc.

**Comment by elong0527** [comment]
@kkmann , I also think we should have most conversation in Github, unless we consider some conversation should go private. 

I can give a try for item 2 by summarizing what we already have. 

How about we build the structure of the functions then we can allocate works using the github issues. I am able to migrate some existing code into the structure. (I believe we already have most metrics except a few complicate metrics)

The people who will lead the complicate metrics can be discussed during our meeting next week.

**Comment by dgkf** [comment]
@kkmann, @elong0527 - How about we make a project board for metrics, where we can file one issue per metric to track the initial development and create a backlog of metrics for us to self-assign and build out? I've used this technique for other modular projects and it's worked pretty well. 

Once an initial function or code base is developed we can migrate to more bug-oriented issues, but this style is helpful for the initial development piece. 

I can take a crack at some onboarding material. @kkmann, is there any reason you prefer to use `index.Rmd` over `README.Rmd`? My understanding is that `pkgdown` will use `README.Rmd` if available. By using it we can try to de-duplicate the on-boarding material as much as possible.

**Comment by elong0527** [comment]
@dgkf  and @kkmann , I always learned great tools from you! I keep wondering how could I set up agile project in github, and now I get it. Yeah, let's go agile with the project board 💯 .

**Comment by kkmann** [comment]
Well, I mainly prefer the index.Rmd since it is specific to the pkdown website, doesn't really matter at this point though ;) I don't really think we need that much project management / boilerplate code atm. I agree that most stuff is fairly simple and it is mainly about finding trustworthy sources for the information we are looking for. 

To me, the most pressing problem is coming up with a clear (initial) scope/structure for the package.
I.e., PR #12 is a good idea if we want to focus on batch processing of sets of packages based on scraping external information (websites). An alternative approach could be built around the package (as tar.gz file), i.e., our functions would be pointed to the tar.gz file directly. This makes a lot of sense to me since the package maintainer should best know where to point us and just parsing the DESCRIPTION file gives a wealth of information already (website, github etc.). Working package-based also had the advantage that we could check the code coverage statistics/test/build without relying on external services. After all, proper validation should be conducted on the user system. Maybe it would make even more sense to validate a package **after** installation? The only real advantage of vectorization is speed and we can safely ignore that at the moment...
I think now that the technical basis is there - thx @elong0527 - we should put a few more days into getting a clearer picture here, right?

**Comment by kkmann** [comment]
Let me clarify:
I guess there are (at least) three distinct approaches here:

1. Focus on collecting package metainformation, primarily by web scraping of relevant and trustworthy resources, along the lines of the `packagemetrics` package
    - can only ever achieve qualification of a package (which seems fine as a first step!)
    - output would essentially be a tibble of individual package metrics
    - main input would be a vector of package names (+ maybe versions?) 
    - primarily geared towards CRAN packages
    - difficult to extend to non-CRAN packages, and there might be viable cases for packages that should not go on CRAN (custom report templates etc.)

2. Running diagnostics on a package .tar.gz directly. I.e. take the sources, build the package, parse the DESCRIPTION, then look for it on CRAN/GitHub/Travis/CodeCov etc. 
    - could be an extension to 1. (right?) in which case it is pretty clear that we should start there
    - allows us to rebuild, re-check, re-analyse code coverage locally not relying on the (potentially ill implemented or non-existent) travis scripts.
    - might still return a simple tibble, but the test etc. output might be so rich that a package individual 'validation report' might be the better option (just sign and be happy ;))

3. Validate the *installation* of an R package. This is slightly different from 2. in that it would run tests using the actually installed version of a package.
    - I put the least thought on this but I feel it would ultimately be the best option since this is what is required in order to validate a package properly, right (1./2. only qualify a package since they are independent from the final execution environment)?
   - should also produce a 'just sign here' validation report

Laying it out like this, I feel 1. is definitely the way to start since it would enter both 2. and 3. Actually, it seems like 1. -> 2. -> 3. would be a nice long-term set of goals, right?

So should we make 1. the first milestone? If so, before starting to implement more stuff we should fix the general layout of the package. I.e. naming conventions for the functions. I do not think we will need to use S3 methods at any point since the input will always be a character (either package name or path to .tar.gz file) I.e., we could define individual metrics to follow the convention metric_[name] where [name] should not contain more than two separate words and no more than xyz characters.

**Comment by elong0527** [comment]
I think our work last month are mostly on step 1. And `packagemetrics` already handle them well for current version of R packages. 

I hope we can use S3 and keep this R package small and sufficient. 

For the first step, one gap is to get the information for archived R packages. 

I would suggest to focus on step 1 and 2. Step 3 rely on the system environment and the IT department of an organization should have a qualification process around its R platform.

**Comment by kkmann** [comment]
So, in general, 1 -> 2 [-> 3] seems like a potential,high-hevel strategy, agreed?
I think we should only probe the standard/most trustworthy sources of information - if something is missing that is also valuable info. I personally would be sceptical about using archived packages since they tend to not be actively maintained.

If we agree on starting with 1., I think we should move the particulars of the naming conventions to a separate issue #16.

**Comment by elong0527** [comment]
I agree to start with 1 and let's discuss in #16 for name conversion.  

The reason to have metrics of all versions of R packages on CRAN is to achieve reproducibility. E.g. a project completed one year ago and start an inspection today should use the same version of R packages at that time to reproduce the result. And we need to have the metrics to support the qualification of those R packages in the version one year ago (which is archived now).

**Comment by elong0527** [comment]
Summary of the meeting for 04/24/2019

* Doug will create a pull request to demo the general function layout
* After package layout defined, We will focous on package metainformation and running diagnostics.
* We need to further evaluate if S3 method is the way to implement this R pacakge. 
* Communicate back to the larger group regarding if a suggest plan for validate the installation
of an R package is within the scope of RValidatioHub Workshop.

**Comment by kkmann** [comment]
#23 is an attempt to build around the idea of computing package metrics based on a locally available .tar.gz file (which seems to be the most robust definition of what a package actually is x)).
We would then parse information from the DESCRIPTION (i.e. is there a github repo?), run code coverage, tests, and checks locally (i.e., prove that the package is working in the context of the already installed R system). Last step would be parsing information form CRAN/github as in package metrics.

**Comment by kkmann** [comment]
@dgkf @elong0527 we still have a few issues regarding bioconductor open - I have very little experience with BC. How would you feel about trying to get a 0.1.0 out without BC? Seems to be mostly a technical challenge of organizing the packages and making sure everything harmonizes but conceptually it shouldn't require that much adjustment, right?

**Comment by elong0527** [comment]
@dgkf , I would like to revisit this issue to ensure we can close it.

**Comment by elong0527** [comment]
We are OK with current structure. Happy to open new issues for the discussion of specific items.
